Endpoint for listing allowed actions on Multiple Resources

darahayes opened the pull request:
This new endpoint allows us to list the actions a user can perform on a list of resources

### Summary
When calling, the endpoint looks like this:

```
GET /authorization/list/some-user-id?resources=foo&resources=bar&resources=baz
```

which will return a list like this:

```
 [
  {
    resource: 'foo',
    actions: ['Read']
  },
  {
    resource: 'bar',
    actions: ['*']
  },
  {
    resource: 'baz',
    actions: []
  }
]
```

__Small Note on Validation__
The endpoint will still work if only one resource is passed in the querystring.

The Joi schema for resources uses ['Array.single()`](https://github.com/hapijs/joi/blob/v10.6.0/API.md#arraysingleenabled) which will optionally accept a single value and put it in an array. e.g.

```
GET /authorization/list/some-user-id?resources=foo
```

will result in `request.query.resources === ['foo']`

### How to Test

Assuming a fresh install, load in the test data with

```
npm run pg:init-test-db
```

and then we can test the endpoint using some of the data inserted by the `fixtures.sql` script with the following:

```
curl -X GET --header 'Accept: application/json' --header 'authorization: ROOTid' --header 'org: WONKA' 'http://localhost:8080/authorization/list/ManyPoliciesId?resources=/myapp/users/dara&resources=/myapp/documents/no_access&resources=/myapp/teams/*'
```

which will give back:

```
[
  {
    "resource": "/myapp/users/dara",
    "actions": [
      "Read"
    ]
  },
  {
    "resource": "/myapp/documents/no_access",
    "actions": []
  },
  {
    "resource": "/myapp/teams/*",
    "actions": [
      "Read",
      "Delete",
      "Edit"
    ]
  }
]
```

darahayes commented:
This should close once merged https://github.com/nearform/udaru/issues/390

darahayes commented:
@p16 yeah come to think of it, the core method is a bit lazy. It's going to list all policies for every resource. Bad...

I'll should be able to reimplement the core method fairly quickly tomorrow morning

paolochiodi commented:
I'd like to report here, for the records, the concern that I had about the list endpoint since we first started discussing it at the beginning of the project.

I find the endpoint as currently implemented (either before or after the PR) a bit unuseful and misleading, because it's not actually returning the list of actions that can be performed, but rather the list of action "rules" that have been inserted for the user.

In the example reported in this PR:
```js
 [
  {
    resource: 'foo',
    actions: ['Read']
  },
  {
    resource: 'bar',
    actions: ['*']
  },
  {
    resource: 'baz',
    actions: []
  }
]
```

we are not telling which actions the user can perform on resource bar, we are just telling that it can do everything. The problem is more visible if we return something like 

```
actions: [
 "authorization:teams:*"
]
```

In these cases the clients needs to re-implment a part of udaru (the pattern matching) in order to get the answer it wants.

I'd suggest to also require a list of actions as input, we could then return a matrix with allow/deny for each resource/action combination. 
Again, following the previous example:

```
GET /authorization/list/some-user-id?resources=foo&resources=bar&resources=ba&actions=Read&actions=Write
```

would result in

```json
{
  "foo": {
    "Read" : "allow",
    "Write": "deny"
  },
  "bar": {
    "Read" : "allow",
    "Write": "allow"
  },
  "baz": {
    "Read" : "deny",
    "Write": "deny"
  }
}
```

I find this solution easier to comprehend and to communicate to the client while still covering 99% of the use cases of the list endpoint

feugy commented:
@paolochiodi, considering how we used the two main endpoints (access and list) in our project, your suggestion would be more a refinement of access than list.

Our UI is using list endpoint to list all possible action on a given resource, without prior knowledge of the action list.
Our backend is using access endpoint to check whether a given action is allowed on given resource.

What you're suggesting is really promising for us, because we often call access 3/4 times with different actions and resources during the same operation. This would be a significant improvement.

However I still think there's a room for an endpoint listing allowed actions without knowing them in advance.

darahayes commented:
Very interesting and valid points guys. Based on what I'm reading here, I believe the functionality @paolochiodi says he would like to see is more of an extension of the access functionality as opposed to the list functionality.

In the original issue #390 we mentioned doing two endpoints. One that provides the functionality in this PR and another almost as Paolo described, although the idea wasn't fleshed out very well. Based off both of your comments there is definitely scope for another endpoint. I can get this done in a separate PR.

darahayes commented:
Okay @p16 I've updated moved the implementation into the `iam.js` script. This has made the implementation about 3-4 times faster than the previous one.

Two things to note:

1. The code is not as elegant as the other `actions` function. I found that chaining a bunch of `filter` and `map` functions made it pretty slow. Instead I wrote it in a more elementary style. I hope it's still readable.
2. I found that there's no need to actually check `isAuthorized({action, resource})` because you already know that the action and resource is inside a statement with `Effect: Allow` If we know `Effect === 'Allow'` then it must be authorized right? Please correct me if I'm wrong but I seem to be getting the desired results.

### Benchmarking
I also added benchmarking tests for the original list endpoint and the new one. 

The new endpoint is about 3-4 times faster than my first implementation. Also based off the tests I've done, the new multi resource endpoint is as fast (and sometimes faster) than the original single list endpoint. (10 resources vs 1 resource).

Please let me know what you think.

### Old Implementation of Multi Resource List Endpoint

```
Stat         Avg     Stdev   Max
Latency (ms) 8718.61 728.34  9975
Req/Sec      10.3    16.02   42
Bytes/Sec    9.11 kB 14.2 kB 38.9 kB

103 requests in 10s, 90.5 kB read
397 errors (394 timeouts)
Detailed Result:

 { title: 'GET authorization/list/{userId}',
  url: 'http://localhost:8080/authorization/list/ManyPoliciesId?resources=/myapp/users/filippo&resources=/myapp/documents/no_access&resources=/myapp/teams/foo&resources=/myapp/teams/foo1&resources=/myapp/teams/foo2&resources=/myapp/teams/foo3&resources=/myapp/teams/foo4&resources=/myapp/teams/foo5&resources=/myapp/teams/foo6&resources=/myapp/teams/foo7',
  requests:
   { average: 10.3,
     mean: 10.3,
     stddev: 16.02,
     min: 0,
     max: 42,
     total: 103,
     sent: 1000 },
  latency:
   { average: 8718.61,
     mean: 8718.61,
     stddev: 728.34,
     min: 7449,
     max: 9975,
     p50: 8719,
     p75: 9311,
     p90: 9731,
     p99: 9967,
     p999: 9975,
     p9999: 9975,
     p99999: 9975 },
  throughput:
   { average: 9113.6,
     mean: 9113.6,
     stddev: 14223.66,
     min: 0,
     max: 38911,
     total: 90537 },
  errors: 397,
  timeouts: 394,
  duration: 10,
  start: 2017-08-10T14:42:55.442Z,
  finish: 2017-08-10T14:43:05.763Z,
  connections: 500,
  pipelining: 1,
  non2xx: 0,
  '1xx': 0,
  '2xx': 103,
  '3xx': 0,
  '4xx': 0,
  '5xx': 0 }
```
### New Implementation of Multi Resource List Endpoint

Whereas with the newer implementation I got the following:

```
Stat         Avg     Stdev   Max
Latency (ms) 4685.39 563.64  5361
Req/Sec      92.4    102.96  276
Bytes/Sec    80.8 kB 90.2 kB 246 kB

924 requests in 10s, 812 kB read
Detailed Result:

 { title: 'GET authorization/list/{userId}',
  url: 'http://localhost:8080/authorization/list/ManyPoliciesId?resources=/myapp/users/filippo&resources=/myapp/documents/no_access&resources=/myapp/teams/foo&resources=/myapp/teams/foo1&resources=/myapp/teams/foo2&resources=/myapp/teams/foo3&resources=/myapp/teams/foo4&resources=/myapp/teams/foo5&resources=/myapp/teams/foo6&resources=/myapp/teams/foo7',
  requests:
   { average: 92.4,
     mean: 92.4,
     stddev: 102.96,
     min: 0,
     max: 276,
     total: 924,
     sent: 1424 },
  latency:
   { average: 4685.39,
     mean: 4685.39,
     stddev: 563.64,
     min: 2232,
     max: 5361,
     p50: 5012,
     p75: 5086,
     p90: 5128,
     p99: 5327,
     p999: 5358,
     p9999: 5361,
     p99999: 5361 },
  throughput:
   { average: 80844.8,
     mean: 80844.8,
     stddev: 90217.89,
     min: 0,
     max: 245759,
     total: 812196 },
  errors: 0,
  timeouts: 0,
  duration: 10,
  start: 2017-08-10T14:43:41.516Z,
  finish: 2017-08-10T14:43:51.679Z,
  connections: 500,
  pipelining: 1,
  non2xx: 0,
  '1xx': 0,
  '2xx': 924,
  '3xx': 0,
  '4xx': 0,
  '5xx': 0 }
```

### Single Resource List Endpoint

```
Running 10s test @ http://localhost:8080/authorization/list/ROOTid/resource_a
500 connections

Stat         Avg     Stdev   Max
Latency (ms) 5267.52 488.04  6196
Req/Sec      61.3    90.38   260
Bytes/Sec    14.8 kB 21.7 kB 63.5 kB

613 requests in 10s, 148 kB read
83 errors (0 timeouts)
Detailed Result:

 { title: 'GET authorization/list/{userId}/{resource}',
  url: 'http://localhost:8080/authorization/list/ROOTid/resource_a',
  requests:
   { average: 61.3,
     mean: 61.3,
     stddev: 90.38,
     min: 0,
     max: 260,
     total: 613,
     sent: 1196 },
  latency:
   { average: 5267.52,
     mean: 5267.52,
     stddev: 488.04,
     min: 4273,
     max: 6196,
     p50: 5333,
     p75: 5560,
     p90: 5923,
     p99: 6179,
     p999: 6193,
     p9999: 6196,
     p99999: 6196 },
  throughput:
   { average: 14771.2,
     mean: 14771.2,
     stddev: 21735.61,
     min: 0,
     max: 63487,
     total: 147733 },
  errors: 83,
  timeouts: 0,
  duration: 10,
  start: 2017-08-10T15:06:28.755Z,
  finish: 2017-08-10T15:06:38.931Z,
  connections: 500,
  pipelining: 1,
  non2xx: 0,
  '1xx': 0,
  '2xx': 613,
  '3xx': 0,
  '4xx': 0,
  '5xx': 0 }
```

p16 commented:
@darahayes 

> I found that there's no need to actually check isAuthorized({action, resource}) because you already know that the action and resource is inside a statement with Effect: Allow If we know Effect === 'Allow' then it must be authorized right? Please correct me if I'm wrong but I seem to be getting the desired results.

The problem is that you may have a `Deny` that will invalidate all the other `Allow`s :( That is why we were using `isAuthorized`. Basically the logic is: as soon as you have `Deny` that matches the action on the resource is denied.

darahayes commented:
@p16 Thanks for the clarification. I'll update accordingly

darahayes commented:
@p16 I think this PR is ready top be merged now. What do you think?

dberesford commented on lib/core/lib/ops/iam.js:
```diff
@@ -42,8 +42,48 @@ module.exports = function (policies) {
+              if (pbac.conditions.StringLike(r, resource)) {
+                statement.Action.forEach((action) => {
+                  isAuthorized({resource, action}, (err, access) => {
+                    if (err) return
```
Should this error be swallowed?

darahayes replied:
@dberesford Yes I wasn't too happy about that piece there. However if you look at the isAuthorized function which is declared in the same file:

```
  function isAuthorized (params, done) {
    return done(null, pbac.evaluate(params))
  }
```

You can see that the `err` can never actually come back. I guess I had that return statement in there to avoid linting issues. What do you think?

dberesford replied:
ah I see, I would refactor that method, probably get rid of it altogether. 

Thanks for making the doc changes!

dberesford commented:
This is great @darahayes!
Could you also add the new endpoints into https://github.com/nearform/udaru/blob/master/docs/overview.md and maybe also work into the trivial example: https://github.com/nearform/udaru/blob/master/docs/example.md
Thanks!

darahayes commented:
@dberesford as requested I've added example documentation

darahayes commented:
@dberesford I've refactored the `isUserAuthorized` function as I saw fit. It's just straight up synchronous now. No point in wrapping a synchronous function that only ever returns true/false with a callback IMO.
